builtin/assign_osh.py

```diff
@@ -203,28 +204,85 @@ def _AssignVarForBuiltin(mem, rval, pair, which_scopes, flags):
     Used for assignment builtins, (( a = b )), {fd}>out, ${x=}, etc.
     """
     lval = LeftName(pair.var_name, pair.blame_word)
-    if pair.plus_eq:
+
+    initializer = None  # type: value.InitializerList
+    if rval is None:
+        # When 'export e+=', then rval is value.Str('')
+        # When 'export foo', the pair.plus_eq flag is false.
+        # Thus, when rval is None, plus_eq cannot be True.
+        assert not pair.plus_eq, pair.plus_eq
+        # NOTE: when rval is None, only flags are changed
+        val = None  # type: value_t
+    elif rval.tag() == value_e.InitializerList:
+        old_val = sh_expr_eval.OldValue(
+            lval,
+            mem,
+            None,  # ignore set -u
+            pair.blame_word)
+        initializer = cast(value.InitializerList, rval)
+
+        # OSH compatibility: "declare -A assoc=(1 2)" is a usage
+        # error.  This code will soon be removed when the
+        # initialization of the form "assoc=(key value)" is supported.
+        if (flag_A and len(initializer.assigns) > 0 and
+                initializer.assigns[0].key is None):
+            e_usage("Can't initialize assoc array with indexed array",
+                    pair.blame_word)
+
+        val = old_val
+        if flag_a:
+            if old_val.tag() == value_e.BashAssoc:
+                # Note: OSH allows overwriting an existing BashArray with an
+                #   empty BashArray.  Bash does not allow this.
+                val = bash_impl.BashArray_New()
+            elif old_val.tag() in (value_e.Undef, value_e.Str,
+                                   value_e.BashArray):
+                # We do not need adjustemnts for -a.  These types are
+                # consistently handled within ListInitialize
+                pass
+            else:
+                e_die(
+                    "Can't convert type %s into BashArray" %
+                    ui.ValType(old_val), pair.blame_word)
+        elif flag_A:
+            if old_val.tag() in (value_e.Undef, value_e.Str,
+                                 value_e.BashArray):
+                # Note: We explicitly initialize BashAssoc for Undef.
+                # Note: OSH allows overwriting an existing BashArray with an
```

Comment:
Arguably, we should not allow this?    Because then someone could write a working OSH script, and then it doesn't work in bash

That is already true in many places probably, but arguably BashArray and BashAssoc should try to be identical


```
$ bash -c 'declare -a a=(a b); declare -A a'
bash: line 1: declare: a: cannot convert indexed to associative array
```

Reply:
#2285